Starbase Cost revamps

Bertie690 opened the pull request:
The code is a bit complex, but it basically withdraws from a global tally to reduce the new item's price down to an initial floor (30% original), before applying an additional category-specific discount to reduce the price  even further (down to a minimum of 20% original).
Also fixed several bugs like incorrect miniaturization rounding & space dock not repairing enough

Fixes #279, #258, #226, #50

TODO:
- [x] Add ISB discount to functions `StarbaseUpgradeCost()` & `ShipDesignCost()`
- [x] Add tests for ISB discount in `StarbaseUpgradeCost()`
- [x] Add tests for `ShipDesignCost()` function
     - Only really need 4-5 tests: 2 for checking ship/starbase costs, 1 for ISB & the rest for error handling in event of invalid hulls/parts 
- [x] Add error checking to `producer.go` & `shipdesign.go`

Bertie690 committed:
Update message.go

Bertie690 committed:
Fixed typos in message.go

Still need to check other files with messages

Bertie690 committed:
Minor typo fixes

Bertie690 committed:
Minor typo fixes

adding missing periods, punctuation fixes, etc.

Bertie690 committed:
Merge branch 'sirgwain:develop' into develop

Bertie690 committed:
Update LRT Descriptions

Minor tweaks & clarifications

Bertie690 committed:
Update LRTsDescriptions.svelte

Typo fix

Bertie690 committed:
Update PRT Descriptions

Effectively resolving my recent issue by myself by updating the PRT description text to include additional info & details

Bertie690 committed:
Update LRTsDescriptions.svelte

Slight additional typo fixes

Bertie690 committed:
Update CE description

forgot to mention that it increases starting prop lvl by 1

Bertie690 committed:
Planet message typofixes

Minor updates (adding missing periods, etc.)

Bertie690 committed:
typo fixes

also made starbase ship build notif distinguish between singular/plural (hopefully shouldn't bork)

Bertie690 committed:
minor message tweaks

added #s for amount of ships involved to battle message descriptions (hopefully won't bork)

Bertie690 committed:
Update PlayerMessageDetail.svelte

minor tweaks

Bertie690 committed:
Update message.go

Bertie690 committed:
minor bugfix to traits.go

fixed a few bugs & starting pop errors

Bertie690 committed:
Improved capacitor/beam deflector code; added tachyon detector cap

Updated caps/deflectors to be multiplicative instead of additive; capped capacitor strength at 2.55x; capped tachyon detector cloak value at 81%

Bertie690 committed:
minor fixes

fixed an oopsie

Bertie690 committed:
Update traits.go

Added minimum decay stat to PRT stats

Bertie690 committed:
Revert "Update traits.go"

This reverts commit 35bdc9de71b8a13129fd6b3334e5abc15e6b59d6.

Bertie690 committed:
minor update to mineralpacket.go (more will come later when I finish stuff)

Bertie690 committed:
Update traits.go

fixed typo that undercounted PP permaform chance by ~20x

Bertie690 committed:
updated packet files

still gotta add the test func

Bertie690 committed:
Torp accuracy rebranding, packet test function

DO NOT MERGE THIS PLS - ONLY MERGE AFTERWARDS TO MAKE SURE I DIDN'T BORK ANYTHING

Bertie690 committed:
Brought over changes from packet branch

just syncing the 2 so I don't go crazy

Bertie690 committed:
Merge branch 'develop' of https://github.com/Bertie690/craig-stars into develop

Bertie690 committed:
Revert "Merge branch 'develop' of https://github.com/Bertie690/craig-stars into develop"

This reverts commit 38ee36323d01f1e56888a1e84cf64a8ae52de721, reversing
changes made to 1277f41db155b848d0b5ad09f9e2cea514c1d91f.

Bertie690 committed:
Revert the revert of the revert

inception babbyyyyy
also i very much hope it works now

Bertie690 committed:
Merge remote-tracking branch 'upstream/develop' into develop

Bertie690 committed:
Merge remote-tracking branch 'upstream/develop' into develop

Bertie690 committed:
Merge branch 'sirgwain:develop' into develop

Bertie690 committed:
Update traits.go

Added IT gate cost decrease

Bertie690 committed:
Starbase Cost Improvements

- Re-wrote function for starbase upgrade costs from the ground up (as well as a new function to calculate ship/starbase design costs instead of relying on `ComputeShipDesignSpec` all the time)
- Passed required values to function calls in ai-related files (since those come with a copy of the game)
- Renamed `torpedoInnacuracyMulti` to `TorpedoBonus` for consistency
- Added techCostOffset.Stargate to IT traits
- Added (currently blank) Tags field to `tech.go`

TODO:
- costcalculator.test.go
- Figure out a way to pass `rules`, `raceSpec` and `TechLevels` objects to every single function that requires them (namely `server/planets.go` & `cs/production.go`
- Add tags to every single item in `techStore.go` and update cost calcs to base it off of those instead

Bertie690 committed:
Cost testing for first function

The code is a bit complex, but it basically withdraws from a global tally to reduce the new item's price down to an initial floor (30% original), before applying an additional category-specific discount to reduce the price  even further (down to a minimum of 20% original).
YES I KNOW THE TESTS ARE FAILING. I had to comment out a line in `cs\production.go` that called the function because it was preventing me from running tests in the first place. I still dunno how to pass it a copy of the rules
Also fixed several bugs like incorrect miniaturization rounding & space dock not repairing enough

Bertie690 committed:
Merge pull request #1 from Bertie690/Bertie690-patch-1

because for some reason my local development instance REFUSES to play nice

Bertie690 committed:
Test Improvements

Bertie690 committed:
Fixed Test Function

It finally works now (yaaaaaaaay)
Still needa add test for the reworked design cost function (which admittedly is very simple and uncontroversial)

Bertie690 committed:
Update production.go

Added rules as part of producer

Bertie690 committed:
Update production.go

Bertie690 committed:
Fixed newProduce function calls to pass rules

Bertie690 committed:
Half did some error handling (still needs more work)

Bertie690 committed:
Update shipdesign.go

Fixed tpyo

Bertie690 committed:
Update production.go

Added error logging for functions

Bertie690 committed:
Update shipdesign.go

Removed ISB starbase discount (gonna add it back in costCalculator.go)

Bertie690 committed:
Added ISB cost discount

`cost.DivideByInt(int(1000*float64(rules.StarbaseComponentCostReduction)/raceSpec.StarbaseCostFactor)`

Bertie690 committed:
Add float rounding to calcs

Reduces risk of conversion errors and adds slightly more accuracy

Bertie690 committed:
Update costcalculator.go

Actually added ISB discount & made component cost reduction apply to staircases only

sirgwain commented on ai/producer.go:
```diff
@@ -325,23 +325,22 @@ func (ai *aiPlayer) getYearsToBuildStarbase(planet *cs.Planet, design *cs.ShipDe
+		cost = design.Spec.Cost
 	}
+	if err != nil {
+		return math.MaxInt, fmt.Errorf("calculate starbase cost %w", err)
```
if you have an err, you return 0, nil, or whatever the default value is for a type. It's expected the caller will not use that value because an error occurred.

Bertie690 replied:
Yeah I should probably change this to reference `GetDesignCost()` directly (rather than calling it indirectly) since I can process the error as well.
Also for the record this is returning MaxInt as it's supposed to return the time needed to build something (and if an error occurs it will never get built anyways)

sirgwain commented on cs/battleweapon.go:
```diff
@@ -44,7 +44,7 @@ type battleWeaponSlot struct {
 
 	// the accuracy bonus to the torpedo
-	torpedoBonus float64
+	TorpedoBonus float64
```
I'm not sure why this was changed to be an exported field? can you revert this please?

Bertie690 replied:
oopsie - might be a leftover from an accidental find and replace some time ago

sirgwain commented on cs/bomb.go:
```diff
@@ -328,7 +328,7 @@ func (b *bomb) retroBombPlanet(planet *Planet, defender *Player, attacker *Playe
 }
 
-// getUnterraformAmount gets the amount we should unterraform with retro bombs
+// Get the amount we should unterraform with retro bombs
```
It's not a hard and fast rule, but go convention is to start a comment with the function name. You can see that in the examples here: https://tip.golang.org/doc/comment

I have mixed feelings on this, which is why I don't apply it everywhere (also I'm just inconsistent)

Bertie690 replied:
okee will undo later

sirgwain commented on db/schema/games/0001_schema.up.sql:
```diff
@@ -392,7 +392,7 @@ CREATE TABLE IF NOT EXISTS techEngines (
   armor INTEGER,
   shield INTEGER,
-  torpedoBonus REAL,
+  TorpedoBonus REAL,
```
definitely revert changes to the schema.

sirgwain commented on cs/techstore.go:
```diff
@@ -1436,24 +1436,24 @@ var UltraStealthCloak = TechHullComponent{Tech: NewTech("Ultra-Stealth Cloak", N
-	HullSlotType:    HullSlotTypeElectrical,
+	Mass:                   1,
+	InitiativeBonus:        2,
+	TorpedoBonus:           .7,
```
this needs to be reverted. The bonus is 30%

sirgwain commented on cs/techstore.go:
```diff
@@ -1436,24 +1436,24 @@ var UltraStealthCloak = TechHullComponent{Tech: NewTech("Ultra-Stealth Cloak", N
-	HullSlotType:    HullSlotTypeElectrical,
+	Mass:                   1,
+	InitiativeBonus:        1,
+	TorpedoBonus:           .8,
```
This needs to be reverted. The bonus is 20%

sirgwain commented on cs/tech.go:
```diff
@@ -100,7 +101,7 @@ type TechHullComponent struct {
 	Armor                     int           `json:"armor,omitempty"`
 	Shield                    int           `json:"shield,omitempty"`
-	TorpedoBonus              float64       `json:"torpedoBonus,omitempty"`
+	TorpedoBonus              float64       `json:"torpedoInaccuracyMulti,omitempty"`
```
this needs to be reverted

sirgwain commented on cs/math.go:
```diff
@@ -13,10 +13,16 @@ func roundToNearest100(value int) int {
+	return math.Round(val*ratio) / ratio
 }
 
+func roundHalfDown(x float64) float64 {
+	if x > 0 {
+		return math.Floor(x + 0.5)
+	}
+	return math.Ceil(x - 0.5)
+}
```
Can you add a couple tests for this function? I'm not sure what its purpose is.

Bertie690 replied:
it’s basically just math.Round() but we round halves down instead of away from 0.
Will make tests at home bc vscode web version can’t use extensions

sirgwain replied:
Are we using this in a scenario where we need -.5 to round to 0?

Bertie690 replied:
Exactly - the easiest example I can think of is a max con B-17 bomber. 25% of 150 resources works out to be 37.5, which rounds DOWN to 37 resources.

sirgwain committed:
Added ResourceType type for Ironium, Germanium, Colonists, Resources, etc.
Removed cost.Max branch where it compared only one costType. It wasn't used and should be a separate function anyway.
Fixed a few syntax errors with missing braces

sirgwain committed:
Merge branch 'develop' of github.com:sirgwain/craig-stars into StarbaseCosts

sirgwain committed:
added unit tests for cost

Bertie690 committed:
Fixed producer.go error handling

Whatchamacallit

Bertie690 committed:
fixed tpyo

Bertie690 committed:
Typofixed battleweapon.go

Gonna switch to vscode.dev at this point lmfao

Bertie690 committed:
fixed things and also improved MT part ratings while I was at it

sirgwain commented on cs/battletoken.go:
```diff
@@ -84,7 +84,7 @@ func newBattleToken(num int, position BattleVector, cargoMass int, token *ShipTo
 		for _, slot := range token.design.Spec.WeaponSlots {
 			weapon := techFinder.GetHullComponent(slot.HullComponent)
-			bws := newBattleWeaponSlot(&battleToken, slot, weapon, hull.RangeBonus, token.design.Spec.TorpedoBonus, token.design.Spec.BeamBonus)
+			bws := newBattleWeaponSlot(&battleToken, slot, weapon, hull.RangeBonus, token.design.Spec.torpedoBonus, token.design.Spec.BeamBonus)
```
heh, this one should be Capitalized because it's an exported field used by the UI and possibly the AI.

Bertie690 replied:
I realize that now - it borking tests.

sirgwain commented on cs/production.go:
```diff
@@ -173,22 +175,46 @@ type builtShip struct {
-			cost = costCalculator.StarbaseUpgradeCost(planet.Starbase.Tokens[0].design, item.design)
+			cost, err = costCalculator.StarbaseUpgradeCost(p.rules, p.player.TechLevels, p.player.Race.Spec, planet.Starbase.Tokens[0].design, item.design)
+			if err != nil {
+				log.Warn().
```
these errors should be returned. producer will need to be updated to return an error. I'm actually surprised it doesn't already since it looks up designs. I'm probably missing a check somewhere. :)

sirgwain commented on cs/shipdesign.go:
```diff
@@ -278,6 +281,16 @@ func ComputeShipDesignSpec(rules *Rules, techLevels TechLevel, raceSpec RaceSpec
+	var err error
+	spec.Cost, err = c.GetDesignCost(rules, techLevels, raceSpec, design)
+	if err != nil {
+		log.Warn().
```
ComputeShipDesignSpec should be updated to return an error instead of logging a warning.

sirgwain commented on cs/tech.go:
```diff
@@ -415,13 +418,16 @@ func (t *Tech) GetPlayerCost(techLevels TechLevel, spec MiniaturizationSpec, cos
 		cost = cost.MultiplyFloat64(1 + costOffset.Bomb)
 	case TechCategoryTorpedo:
 		cost = cost.MultiplyFloat64(1 + costOffset.Torpedo)
+		//case TechCategoryOrbital && {some way to check if item is gate}
+		//	cost = cost.MultiplyFloat64(1 + costOffset.Stargate)
```
was this left in by mistake?

Bertie690 replied:
Nah - I just wanted to put a comment there to remind myself to fix thing once the tag system gets implemented. 
Probably needs a big orange "TODO" marker come to think about it...

sirgwain committed:
using TechCategory.Orbital instead of a string. Reusing results of getHullComponent.
Fixed some warnings (errors are not supposed to start with capital letters.)

sirgwain committed:
fix GetDesignCost GetHull check.
Fixed casing on torpedoBonus

Bertie690 committed:
Update battleweapon.go

Bertie690 committed:
Update bomb.go

sirgwain committed:
fixed GetDesignCost to include slot quantity
Fixed unit tests to have correct data after rounding is working

sirgwain committed:
Merge branch 'StarbaseCosts' of github.com:Bertie690/craig-stars into StarbaseCosts

sirgwain committed:
fix more tests for updated cost calcs

Bertie690 committed:
Update production.go

Made overhead actually work properly by dividing the packet cargo amount by the cost factor

Bertie690 committed:
minor fix production.go

Bertie690 committed:
Added janky stargate cost multiplier

Yes I know it's jank, but it better than nothing ok??

sirgwain committed:
Merge branch 'develop' of github.com:sirgwain/craig-stars into StarbaseCosts

sirgwain committed:
Added error handling/propagation to produce()

sirgwain committed:
setting tech store to wasm rules

sirgwain committed:
Merge branch 'develop' of github.com:sirgwain/craig-stars into StarbaseCosts

Bertie690 committed:
Added error handling in related functions & ISB test case

Also added explicitly failing battle test case for if the components are wrong (if component bad it no work)

Bertie690 committed:
added test func

100% broken needs review
zzzzz

Bertie690 committed:
Fixed test for costCalculate.GetDesignCost

Bertie690 committed:
Added nil hull check to ComputeShipDesignSpec

Bertie690 committed:
Update battle_test.go

sirgwain committed:
Merge branch 'develop' of github.com:sirgwain/craig-stars into StarbaseCosts

sirgwain committed:
fixed tests

sirgwain committed:
changed logging on WithSpec. it's used by tests
